### ISSUE TITLE
TypeDefine issues with formatResponse and ApolloServer

### ISSUE THREAD

**Issue by tracy-liucui** [apollographql/apollo-server#3590]
Currently, I encountered some type errors with `apollo-server` package.

1. For the`formatResponse` typeDefine,  `response` can be null, but the returned type doesn't accept `null`
https://github.com/apollographql/apollo-server/blob/8059326ce43a776b6f34e1a6de4cb78da4e4a2b3/packages/apollo-server-core/src/graphqlOptions.ts#L48-L51

2. Get incompatible error when using custom context type because the `ApolloServer` is not a generic.

Here is the example:
```typescript
interface Context {
  foo: string
}

const formatResponse: GraphQLOptions<Context>['formatResponse'] = (
  response
) => response //<-- 1. error: null is not assignable to 'GraphQLResponse'

const server = new ApolloServer({
  formatResponse, //<-- 2. error: Property 'foo' is missing in type '{}' but required in type 'Context'
})
```

**Comment by Markus-ipse** [comment]
I just updated from 2.25.2 to 3.3.0 (I know I am late to the 3.x party) and now formatResponse no longer complain about possibly returning null.

**Comment by glasser** [comment]
As @Markus-ipse says, the first concern was fixed earlier this year (#5089, v2.23.0).

If the second concern is still an issue, please file a new issue including a full reproduction (eg starting with `git clone` or using codesandbox.io) demonstrating it.
